### DISCUSSION TITLE
Run different websocket framework rather than built-in websocket

### DISCUSSION THREAD

**Discussion by phtran-dev** [fastapi/fastapi#9421] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, WebSocket
from fastapi.websockets import WebSocketDisconnect


app = FastAPI()


@app.websocket('/test')
async def test(websocket: WebSocket, prefix: str):
	await websocket.accept()
	try:
		while True:
			message = await websocket.receive_text()
			await websocket.send_text(prefix+message)
	except WebSocketDisconnect:
		pass
```

### Description

Hi All,
I know that FastAPI has built-in Websocket support. However, this socket is not compatible with my application's requirements. I want to run a different websocket framework (https://github.com/Kitware/wslink) besides FastAPI (in which I can use RestAPI). I wonder if FastAPI can support it ?

Thanks

### Operating System

Linux

### Operating System Details

https://github.com/Kitware/wslink

### FastAPI Version

0.68.0

### Python Version

Python 3.9.5

### Additional Context

Hi All,
I know that FastAPI has built-in Websocket support. However, this socket is not compatible with my application's requirements. I want to run a different websocket framework (https://github.com/Kitware/wslink) besides FastAPI (in which I can use RestAPI). I wonder if FastAPI can support it ?

Thanks

**Comment by Kludex** [comment, marked as answer]
FastAPI is not the one who chooses the `websocket` framework to be used. Using `uvicorn` you can either use `wsproto` or `websockets` packages.

**Reply by phtran-dev** [reply]
Can you elaborate more ? Sanic has feature “hook” which can integrate into the core and customize to work with ? Does Fastapi have that feature ?

**Reply by Kludex** [reply]
> Does Fastapi have that feature ?

No, FastAPI doesn't have the "hook" feature.

> Can you elaborate more ?

Yes. FastAPI follows an interface called [ASGI], and it's in charge of sending messages to the ASGI server (which in this case I imagine is Uvicorn). Uvicorn is the package that effectively handle the communication with the client, and to do that, it can be done with two packages: `websockets` and `wsproto`, but you shouldn't be concerned about what Uvicorn uses, since it follows what is defined in the WebSockets RFCs.

[ASGI]: https://asgi.readthedocs.io/en/latest/
